### PR TITLE
PERF: only output an html id if a style is applied

### LINKED ISSUE #20695: Huge html with df.style.render to due css duplications
When creating html with `df.style` e.g.
```python
df.style.apply(color_f, axis=1).render()
```
pandas assigns to each html cell of the table a unique css class. This means huge html are created, while it would be possible to use the same class for cell with the same style.

<details>

------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.14-300.fc27.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.1.3
pip: 9.0.1
setuptools: 36.0.1
Cython: 0.27.3
numpy: 1.14.2
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 5.4.1
sphinx: None
patsy: 0.4.1
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.11
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by Moisan**
- [x] closes #20695
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Greatly reduces the amount of `id` tag in the cells by only assigning one when a style is applied to that cell. If that is the not correct approach for solving #20695, I am open to suggestions.

Also I was confused by the `%-` tags to handle the whitespaces. Is there a reason the default options to handle whitespace are not used?

### CODE REVIEW

**Comment by TomAugspurger** [comment]
I think this should be a parameter given to the Styler constructor, not something that happens implicitly.

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -65,6 +65,8 @@ class Styler(object):
         a unique identifier to avoid CSS collisions; generated automatically
     caption: str, default None
         caption to attach to the table
+    all_ids: bool, default False
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: Perhaps ``cell_ids`` is a bit more descriptive?

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -65,6 +65,8 @@ class Styler(object):
...
+        if True, each cell will have an ``id`` attribute in their HTML tag.
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: Capitalize `if`.

Could you include the form these take? I believe it's `T_<uuid>_row<i>_col<j>` for values in the table.

File: pandas/io/formats/style.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/io/formats/style.py]: And the default should be True (backwards compatible)

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -307,14 +310,16 @@ def format_attr(pair):
...
-                    "display_value": formatter(value),
-                    "is_visible": (c not in hidden_columns)
-                })
+                row_dict = {"type": "td",
+                            "value": value,
+                            "class": " ".join(cs),
+                            "display_value": formatter(value),
+                            "is_visible": (c not in hidden_columns)}
+                # only add an id if the cell has a style
+                if self.all_ids or \
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: This should just be `if self.cell_ids` I think. We don't want a special case for the number of styles.

File: pandas/io/formats/templates/html.tpl [hunk 4]
```diff
@@ -50,17 +50,21 @@
...
-    <tr> 
-        {%- for c in r %} 
-        {%- if c.is_visible != False %} 
+    {% block before_rows %}{% endblock before_rows %}
+    {% for r in body %}
+    {% block tr scoped %}
+    <tr>
+        {%- for c in r %}
+        {%- if c.is_visible != False %}
+        {%- if c.id: %}
```
Comment (TomAugspurger) [inline pandas/io/formats/templates/html.tpl]: I'm not that familiar with jinja, but presumably there's a way to deduplicate this a bit? I worry about these two cases drifting apart in the future.

File: pandas/io/formats/templates/html.tpl [hunk 5]
```diff
@@ -50,17 +50,17 @@
...
-    <tr> 
-        {%- for c in r %} 
-        {%- if c.is_visible != False %} 
-        <{{ c.type }} id="T_{{ uuid }}{{ c.id }}" class="{{ c.class }}" {{ c.attributes|join(" ") }}>{{ c.display_value }}</{{ c.type }}>
-        {%- endif %} 
-        {%- endfor %} 
-    </tr> 
-    {%- endblock tr %}
+    {% block before_rows %}{% endblock before_rows %}
+    {% for r in body %}
```
Comment (TomAugspurger) [inline pandas/io/formats/templates/html.tpl]: Does removing the opening `-` change the whitespace? Or was it redundant?

Comment (Moisan) [reply pandas/io/formats/templates/html.tpl]: It was redundant.

Comment (Moisan) [reply pandas/io/formats/templates/html.tpl]: I believe it was redundant.

Comment (TomAugspurger) [reply pandas/io/formats/templates/html.tpl]: I think so too.

**Review by TomAugspurger — APPROVED** [review]
Just removed a backslash.

Ping on green.

**Comment by Moisan** [comment]
@TomAugspurger ping

**Pull request merged**